stdlib/path.gr

```diff
@@ -0,0 +1,411 @@
+let rec relativizeDepth = ((up1, s1), (up2, s2)) =>
+  match ((dirsUp(up1), dirsUp(up2), s1, s2)) {
+    (Zero, Zero, [hd1, ...tl1], [hd2, ...tl2]) when hd1 == hd2 =>
+      relativizeDepth((0, tl1), (0, tl2)),
+    (Zero, Zero, [], _) => Ok((up2, s2)),
+    (Zero, Zero, _, _) => Ok((List.length(s1), s2)),
+    (Positive, Positive, _, _) => relativizeDepth((up1 - 1, s1), (up2 - 1, s2)),
+    (Zero, Positive, _, _) => Ok((List.length(s1) + up2, s2)),
+    (Positive, Zero, _, _) => Err(ImpossibleRelativization),
+  }
+
+/**
+ * Attempts to construct a new relative path which will lead to the destination
+ * path from the source path. If the source and destination are both absolute
+ * paths but are on different roots, `Err(Path.DifferentRoots)` will be
+ * returned, and if the route to the destination cannot be concretely
+ * determined from the source, `Err(Path.ImpossibleRelativization)` will be
+ * returned.
+ * 
+ * @param source: The source path to find a path from
+ * @param dest: The destination path to find the path to
+ * @returns An `Ok` variant containing the relative path if successfully parsed or `Err` otherwise
+ * 
+ * @example // relativeTo(/usr, /usr/bin) => ./bin
+ * @example // relativeTo(/home/me, /home/me) => .
+ * @example // relativeTo(/usr/bin, /etc) => ../../etc
+ * @example // relativeTo(.., ../../thing) => ../thing
+ * @example // relativeTo(/usr/bin, C:/Users) => Err(DifferentRoots)
+ * @example // relativeTo(../here, ./there) => Err(ImpossibleRelativization)
+ * 
+ * @since v0.5.4
+ */
+export let relativeTo = (source: Path<a>, dest: Path<a>) => {
+  let result = match ((source, dest)) {
+    ((Abs(r1), s1), (Abs(r2), s2)) =>
+      if (r1 != r2) Err(DifferentRoots)
+      else relativizeDepth((0, List.reverse(s1)), (0, List.reverse(s2))),
+    ((Rel(up1), s1), (Rel(up2), s2)) =>
+      relativizeDepth((up1, List.reverse(s1)), (up2, List.reverse(s2))),
+    _ => fail "",
+  }
+
+  match (result) {
+    Ok((depth, segs)) => Ok((Rel(depth), List.reverse(segs))),
+    Err(err) => Err(err),
+  }: Result<Path<Relative>, RelativizationError>
+}
+
+/**
+ * Returns the path corresponding to the parent directory of the given path.
+ * 
+ * @param path: The path to get the parent of
+ * @returns A path corresponding to the parent directory of the given path
+ * 
+ * @since v0.5.4
+ */
+export let parent = (path: Path<a>) =>
+  match (path) {
+    (base, [_, ...rest]) => (base, rest),
+    (Rel(upDirs), []) => (Rel(upDirs + 1), []),
+    (Abs(_) as base, []) => (base, []),
+  }: Path<a>
+
+/**
+ * Returns the base name (named final segment) of a path.
+ * 
+ * @param path: The path to get the name of
+ * @returns A `Some` variant containing the name of the path, or `None` if the path does not have one
+ * 
+ * @example Path.name(Option.unwrap(Path.relative("./dir/file.txt"))) // Some("file.txt")
+ * @example Path.name(Option.unwrap(Path.relative(".."))) // None
+ * 
+ * @since v0.5.4
+ */
+export let name = (path: Path<a>) =>
```

Comment:
Should we name this `filename` or `basename`? Trying to think inline with other stdlibs

Reply:
`filename` seems to me like it's specifically tailored to paths that point to files rather than directories (which I guess are also files technically but you don't generally think of them that way). I was initially partial to `basename` but usually that is paired with `dirname` for the `parent` operation, implying that a "name" (which to me implies a string) would be returned, whereas `parent` actually returns a `Path` and not a string. I'm coming around more to the idea of `basename` though, I can see how it's clearer than just `name`... Any thoughts about changing or keeping the name `parent`?